core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala

```diff
@@ -163,7 +187,7 @@ class OpenHashSet[@specialized(Long, Int) T: ClassTag](
    *                 to a new position (in the new data array).
    */
   def rehashIfNeeded(k: T, allocateFunc: (Int) => Unit, moveFunc: (Int, Int) => Unit) {
-    if (_size > _growThreshold) {
+    if (_occupied > _growThreshold) {
```

Comment:
I dont see any value in _occupied - on contrary it can cause very bad behavior if there is a lot of remove's expected.
`_size` is a better metric to decide to rehash and grow.

Reply:
When 'remove' is called, '_size' is decremented. But, an entry is not released. This is  a motivation to introduce 'occupied'.
I will try to use another implementation without 'remove' while it may introduce some overhead.

Reply:
There is no explicitly entry here - it is simply unoccupied slots in an array.
The slot is free, it can be used by some other (new) entry when insert is called.

It must be trivial to see how very bad behavior can happen with actual size of set being very small - with a series of add/remove's : resulting in unending growth of the set.

something like this, for example, is enough to cause set to blow to 2B entries:
```
var i = 0
while (i < Int.MaxValue) {
  set.add(1)
  set.remove(1)
  assert (0 == set.size)
  i += 1
}
```

Reply:
For accuracy sake - my example snippet above will fail much earlier - due to OpenHashSet. MAX_CAPACITY. Though that is probably not the point anyway :-)